### PR TITLE
Update javascript renderer

### PR DESCRIPTION
**Pull request by eulertour**
This simplifies the logic in the JS renderer and updates it to work with the newer scene changes such as wait animations and the new skipping logic. Major changes include:

1. `FrameServer` is completely rewritten and no longer distinguishes between Wait and other Animations
2. The protobufs used to communicate with the javascript frontend were also rewritten
3. Added a `__deepcopy__` method to `Scene` with logic to copy updaters
4. Combined `Scene._get_animation_time_progression()` and `Scene._get_wait_time_progression()`
5. Added `Scene.compile_animation_data()` which contains the initialization logic when an animation is played
6. Added `Scene.update_to_time()` so that Scenes can be updated to a time with a function

### CODE REVIEW

**Comment by eulertour** [comment]
Having to format generated files is sort of silly, so feel free to ignore the black check until I get to it.

File: manim/grpc/gen/frameserver_pb2.py [hunk 1]
```diff
@@ -1,743 +1,790 @@
 # -*- coding: utf-8 -*-
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
 # source: frameserver.proto
-
+"""Generated protocol buffer code."""
```
Comment (naveen521kk) [inline manim/grpc/gen/frameserver_pb2.py:4]: This is generated one right? So it doesn't need review?

File: manim/grpc/gen/frameserver_pb2_grpc.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
 # Generated by the gRPC Python protocol compiler plugin. DO NOT EDIT!
+"""Client and server classes corresponding to protobuf-defined services."""
```
Comment (naveen521kk) [inline manim/grpc/gen/frameserver_pb2_grpc.py:2]: Is this also generated one?

File: manim/grpc/gen/renderserver_pb2.py [hunk 3]
```diff
@@ -1,550 +1,272 @@
 # -*- coding: utf-8 -*-
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
 # source: renderserver.proto
-
+"""Generated protocol buffer code."""
```
Comment (naveen521kk) [inline manim/grpc/gen/renderserver_pb2.py:4]: This also generated?

File: scripts/simple_cors_http_server.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+        self.send_header("Access-Control-Allow-Origin", "*")
+        SimpleHTTPRequestHandler.end_headers(self)
+
+
+if __name__ == "__main__":
+    test(
+        CORSRequestHandler,
+        HTTPServer,
+        port=int(sys.argv[1]) if len(sys.argv) > 1 else 8000,
+    )
```
Comment (naveen521kk) [inline scripts/simple_cors_http_server.py:17]: Why is this needed?

File: manim/grpc/update_protos.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-python \
+poetry run python \
```
Comment (naveen521kk) [inline manim/grpc/update_protos.sh:3]: Would suggest reverting this change in this file. I don't think poetry is needed for this.

File: manim/__main__.py [hunk 6]
```diff
@@ -75,26 +76,28 @@ def main():
...
+        else:
+            for SceneClass in scene_classes_from_file(input_file):
+                try:
+                    scene = SceneClass()
+                    scene.render()
+                    open_file_if_needed(scene.renderer.file_writer)
+                except Exception:
+                    print("\n\n")
+                    traceback.print_exc()
+                    print("\n\n")
```
Comment (naveen521kk) [inline manim/__main__.py:100]: This looks weird why? Like why to print two new lines?

File: manim/grpc/impl/frame_server_impl.py [hunk 7]
```diff
@@ -10,250 +10,279 @@
 import subprocess as sp
 import threading
 import time
+import traceback
 import ctypes
```
Comment (naveen521kk) [inline manim/grpc/impl/frame_server_impl.py:14]: Unnecessary Import
```suggestion
```

**Review by naveen521kk — CHANGES_REQUESTED** [review]
Some comments. Also, I didn't look into the files in `gen` folder.

File: manim/__main__.py (hunk 6 above)
Comment (naveen521kk) [inline manim/__main__.py:100]: Why not something like
```suggestion
                except Exception as e:
                    print(e)
```

File: manim/grpc/impl/frame_server_impl.py [hunk 8]
```diff
@@ -10,250 +10,279 @@
...
-                # the same way Scene.play does
-                selected_scene.update_animation_to_time(time)
-                selected_scene.update_frame()
-                serialized_mobject_list, duration = selected_scene.add_frame(
-                    selected_scene.get_frame()
-                )
-                frame_response = list_to_frame_response(
-                    selected_scene, duration, serialized_mobject_list
+        except Exception as e:
+            traceback.print_exc()
```
Comment (naveen521kk) [inline manim/grpc/impl/frame_server_impl.py:165]: Again is this any different that printing `e` directly? Or I'm mistaken?

File: manim/grpc/impl/frame_server_impl.py [hunk 9]
```diff
@@ -10,250 +10,279 @@
...
+import copy
+from ...mobject.value_tracker import ValueTracker
+from ...mobject.types.vectorized_mobject import VMobject
+from ...mobject.types.image_mobject import ImageMobject
 
 
-class FrameServer(frameserver_pb2_grpc.FrameServerServicer):
-    def animation_index_is_cached(self, animation_index):
-        return animation_index < len(self.keyframes)
+class MyEventHandler(FileSystemEventHandler):
```
Comment (naveen521kk) [inline manim/grpc/impl/frame_server_impl.py:32]: Why is it named as `MyEventHandler`? Could be named anything else IMO than having something like this in a library.

File: manim/grpc/impl/frame_server_impl.py [hunk 10]
```diff
@@ -10,250 +10,279 @@
...
+                        frameserver_pb2.Animation(
+                            name=animations_to_name(scene.animations),
+                            duration=scene.duration,
+                        )
+                        for scene in self.keyframes
+                    ],
+                ),
+            )
+        except Exception as e:
+            traceback.print_exc()
```
Comment (naveen521kk) [inline manim/grpc/impl/frame_server_impl.py:182]: Here also. Is this any different than printing `e`?

File: manim/grpc/update_protos.sh (hunk 5 above)
Comment (naveen521kk) [reply manim/grpc/update_protos.sh:3]: Also possibly move to `scripts` folder?

**Review by naveen521kk — CHANGES_REQUESTED** [review]

**Review by naveen521kk — CHANGES_REQUESTED** [review]
Run black :wink:.

File: manim/grpc/gen/frameserver_pb2.py (hunk 1 above)
Comment (eulertour) [reply manim/grpc/gen/frameserver_pb2.py:4]: That's right.

File: manim/grpc/gen/renderserver_pb2.py (hunk 3 above)
Comment (eulertour) [reply manim/grpc/gen/renderserver_pb2.py:4]: Yes

File: scripts/simple_cors_http_server.py (hunk 4 above)
Comment (eulertour) [reply scripts/simple_cors_http_server.py:17]: The renderer needs access to the files used for image mobjects in order to render them but I don't know how to configure it to access local files without [disabling webSecurity](https://stackoverflow.com/questions/48777336/is-disabling-websecurity-on-the-electron-window-the-only-way-to-display-local-im) which is considered bad practice. So I used this to serve the files from a local server. It may not be necessary but I don't know of another safe way to handle ImageMobjects.

Comment (eulertour) [reply scripts/simple_cors_http_server.py:17]: [This](https://github.com/ManimCommunity/manim-renderer/blob/master/src/App.vue#L314) is where it's handled by the renderer.

File: manim/grpc/update_protos.sh (hunk 5 above)
Comment (eulertour) [reply manim/grpc/update_protos.sh:3]: Poetry is needed to access protoc when developing with poetry. I moved the script to scripts/.

File: manim/__main__.py (hunk 6 above)
Comment (eulertour) [reply manim/__main__.py:100]: This is just copying the way it's already done in `__main__.py`.

File: manim/grpc/impl/frame_server_impl.py (hunk 7 above)
Comment (eulertour) [reply manim/grpc/impl/frame_server_impl.py:14]: fixed

File: manim/__main__.py (hunk 6 above)
Comment (eulertour) [reply manim/__main__.py:100]: Doing so would change the format of the error messages, which isn't the point of this PR.

File: manim/grpc/impl/frame_server_impl.py (hunk 8 above)
Comment (eulertour) [reply manim/grpc/impl/frame_server_impl.py:165]: When I just `print(e)` the location of the error isn't printed. I think it's related to the way gRPC runs the server.

Comment (naveen521kk) [reply manim/grpc/impl/frame_server_impl.py:165]: Ok then fine.

File: manim/grpc/impl/frame_server_impl.py (hunk 10 above)
Comment (eulertour) [reply manim/grpc/impl/frame_server_impl.py:182]: When I just print(e) the location of the error isn't printed.

File: manim/grpc/gen/frameserver_pb2_grpc.py (hunk 2 above)
Comment (eulertour) [reply manim/grpc/gen/frameserver_pb2_grpc.py:2]: yes

File: manim/grpc/impl/frame_server_impl.py (hunk 9 above)
Comment (eulertour) [reply manim/grpc/impl/frame_server_impl.py:32]: Noted

**Review by naveen521kk — APPROVED** [review]
LGTM! :tada:

**Comment by kolibril13** [comment]
I did not look too much into the code, but I tested it with the latest artefact from here, https://github.com/ManimCommunity/manim-renderer/actions/runs/431557071, and it works properly.

https://user-images.githubusercontent.com/44469195/102692100-61f11980-4211-11eb-9ba7-85e83048da0b.mp4

**Review by kolibril13 — APPROVED** [review]
